pkg/ccl/changefeedccl/sink_webhook.go

```diff
@@ -311,42 +315,66 @@ func (i *inflightTracker) maybeSetError(err error) {
 }
 
 // StartRequest enqueues one inflight message to be flushed.
-func (i *inflightTracker) StartRequest() {
+func (i *inflightTracker) StartRequest(ctx context.Context, bytes int64) error {
 	i.flushMu.Lock()
 	defer i.flushMu.Unlock()
+	defer i.memMu.Unlock()
 	i.inflightGroup.Add(1)
+	i.memMu.Lock()
+	err := i.memMu.mem.Grow(ctx, bytes)
+	return err
 }
 
 // FinishRequest tells the inflight tracker one message has been delivered.
-func (i *inflightTracker) FinishRequest() {
+func (i *inflightTracker) FinishRequest(ctx context.Context, bytes int64) {
+	i.memMu.Lock()
+	defer i.memMu.Unlock()
+	i.memMu.mem.Shrink(ctx, bytes)
 	i.inflightGroup.Done()
 }
 
 // Wait waits for all inflight messages to be delivered (inflight = 0) and
 // returns a possible error. New messages delivered via EmitRow and
 // EmitResolvedTimestamp will be blocked until this returns.
-func (i *inflightTracker) Wait() error {
+func (i *inflightTracker) Wait(ctx context.Context) error {
 	i.flushMu.Lock()
 	defer i.flushMu.Unlock()
 	i.inflightGroup.Wait()
 	var err error
 	select {
+	case <-ctx.Done():
+		return ctx.Err()
 	case err = <-i.errChan:
 	default:
 	}
 	return err
 }
 
+func (i *inflightTracker) Close(ctx context.Context) {
+	i.memMu.Lock()
+	defer i.memMu.Unlock()
+	i.memMu.mem.Close(ctx)
+	close(i.errChan)
+}
+
 func (s *webhookSink) EmitRow(
 	ctx context.Context, _ TopicDescriptor, key, value []byte, _ hlc.Timestamp,
 ) error {
-	s.inflight.StartRequest()
+	j, err := encodePayloadWebhook(value)
```

Comment:
moved encoding back into EmitRow because the memory monitor needs to know the size of the actual payload before it gets shipped off to the worker